imports/validators.py

```diff
@@ -23,22 +29,61 @@ def __call__(self, value):
             )
 
         with zipfile.ZipFile(value.file) as zfile:
-            required_extensions = ('.shp', '.shx', '.dbf')
-            filenames, extensions = [], []
-            for filename in zfile.namelist():
-                if filename.endswith(required_extensions):
-                    filename, extension = os.path.splitext(filename)
-                    filenames.append(filename)
-                    extensions.append(extension)
-
-            if len(set(filenames)) > 1:
-                raise ValidationError(
-                    self.messages['mismatched_filenames'],
-                    code='mismatched_filenames',
-                )
-
-            if set(extensions) != set(required_extensions):
-                raise ValidationError(
-                    self.messages['missing_required_files'],
-                    code='missing_required_files',
-                )
+            self._validate_files(zfile)
+
+            dbf, shp, shx = sorted([name for name in zfile.namelist() if name.endswith(REQUIRED_SHAPEFILE_EXTENSIONS)])
+            with zfile.open(shp) as shp, zfile.open(shx) as shx, zfile.open(dbf) as dbf:
+                shapefile_reader = shapefile.Reader(shp=shp, shx=shx, dbf=dbf)
+
+            self._validate_fields(shapefile_reader)
+            self._validate_feature_classes(shapefile_reader)
+
+    def _validate_files(self, zfile):
+        filenames, extensions = [], []
+        for filename in zfile.namelist():
+            if filename.endswith(REQUIRED_SHAPEFILE_EXTENSIONS):
+                filename, extension = os.path.splitext(filename)
+                filenames.append(filename)
+                extensions.append(extension)
+
+        if len(set(filenames)) > 1:
```

Comment:
Is type casting really needed in these situations? I have a hard time seeing how this will make anything faster or better, is there any benefit to casting the list to set here? Feels unnessesary by there might be something I'm missing.

Reply:
The filenames are the name prefixes of `.shp`, `.shx` and `.dbf` files, this is to make sure that all the name prefixes are the same for all three files. This is the simplest way I can think of, but suggestions are very welcome.

Reply:
I guess you are right, this is most likey the cleanest way of doing that indeed. Sometimes comments for these things would be of great help, not only to instantly understand what is going on but to also understand why.